benchopt/base.py

```diff
@@ -392,7 +392,11 @@ def set_dataset(self, dataset):
             parameters[key] = getattr(self, key)
         self.set_data(**data)
         for key in self._parameters:
-            if parameters[key] != getattr(self, key):
+            has_changed = (any(parameters[key] != getattr(self, key))
+                           if hasattr(parameters[key], '__iter__')
+                           else parameters[key] != getattr(self, key))
```

Comment:
A more comprehensive implem, which would also work for list.

```suggestion
            if hasattr(parameters[key], '__iter__'):
                has_changed = any(
                    p != a for p, a in zip(parameters[key], getattr(self, key)
                )
            else:
                has_changed = parameters[key] != getattr(self, key)
```

Reply:
This is a buggy implementation,

Indeed,
```python
[1, 2] != [1, 2, 3]
# output: True

any(i != j for i, j in zip([1, 2], [1, 2, 3]))
# output: False

# [(i, j) for i, j in zip([1, 2], [1, 2, 3])] gives [(1, 1), (2, 2)]
```

So your code won't produce the desired behavior.

Reply:
Indeed

Reply:
I made a small change in the implementation.
Hope it's more comprehensible.